haystack_experimental/components/agents/agent.py

```diff
@@ -0,0 +1,264 @@
+
+        if state_schema is not None:
+            _validate_schema(state_schema)
+        self.state_schema = state_schema or {}
+
+        self.model = model
+        self.generation_kwargs = generation_kwargs or {}
+        self.tools = tools or []
+        self.system_prompt = system_prompt
+        self.exit_condition = exit_condition
+        self.api_key = api_key
+
+        component.set_input_type(instance=self, name="messages", type=List[ChatMessage])
+        output_types = {"messages": List[ChatMessage]}
+        for param, config in self.state_schema.items():
+            component.set_input_type(self, name=param, type=config["type"], default=None)
+            output_types[param] = config["type"]
+        component.set_output_types(self, **output_types)
+
+        self._initialize_pipeline()
+
+    def _initialize_pipeline(self) -> None:
+        """Initialize the component pipeline with all necessary components and connections."""
+        provider, model_name = self.model.split(":")
+
+        if provider == "anthropic":
+            anthropic_import.check()
+
+        # Initialize components
+        generator = _PROVIDER_GENERATOR_MAPPING[provider](
+            model=model_name,
+            tools=self.tools,
+            api_key=self.api_key,
+            generation_kwargs=self.generation_kwargs,
+        )
+        joiner = BranchJoiner(type_=List[ChatMessage])
+        tool_invoker = ToolInvoker(tools=self.tools, raise_on_failure=False)
+        context_joiner = BranchJoiner(type_=State)
+
+        # Configure router conditions
+        if self.exit_condition == "text":
+            exit_condition_template = "{{ llm_messages[0].tool_call is none }}"
+        else:
+            exit_condition_template = (
+                "{{ llm_messages[0].tool_call is none or (llm_messages[0].tool_call.tool_name == '"
+                + self.exit_condition
+                + "' and not tool_messages[0].tool_call_result.error) }}"
+            )
+
+        router_output = "{%- set assistant_msg = (llm_messages[0].text|trim or 'Tool:')"\
+                        "|assistant_message(none, none, llm_messages[0].tool_calls) %}"\
+                        "{{ original_messages + [assistant_msg] + tool_messages }}"
+
+        routes = [
+            {
+                "condition": exit_condition_template,
+                "output": router_output,
+                "output_type": List[ChatMessage],
+                "output_name": "exit",
+            },
+            {
+                "condition": "{{ True }}",  # Default route
+                "output": router_output,
+                "output_type": List[ChatMessage],
+                "output_name": "continue",
+            },
+        ]
+
+        router = ConditionalRouter(
+            routes=routes,
+            custom_filters={"assistant_message": ChatMessage.from_assistant},
+            unsafe=True,
+        )
+
+        # Set up pipeline
+        self.pipeline = Pipeline()
```

Comment:
Exposing `max_runs_per_component` in the init of Agent could be nice here especially since we will be doing looping.

Reply:
Yes, I can do that. One thing I noticed about the `max_runs_per_component` while working on the pipeline.run logic is that we raise an exception when we exceed the limit. I'd say that that isn't very useful for an Agent. It would be possible to change the logic and just remove components that reached their limit from the set of components that can be scheduled. This is for another discussion though.